### PR TITLE
[async] [lang] [opt] Add ti.loop_unique(covers=...) to improve task dependence analysis

### PR DESCRIPTION
**Pull request by xumingkuan**
Related issue = #742

Also improves `same_value` and `whole_kernel_cse`.

Usage example:
```python
import taichi as ti

ti.init(arch=ti.cpu, debug=False, async_mode=True, print_preprocessed=False,
        print_ir=False)

N = 16
block = ti.root.pointer(ti.i, N)
x = ti.field(ti.i32)
block.place(x)

block2 = ti.root.pointer(ti.i, N)
y = ti.field(ti.i32)
block3 = block2.dense(ti.i, 2)
block3.place(y)

print('x: ', x.snode.id)
print('y: ', y.snode.id)
print('block: ', block.snode.id)
print('block2: ', block2.snode.id)
print('block3: ', block3.snode.id)

@ti.kernel
def test():
    for i in x:
        p = ti.loop_unique(x[i], covers=block3)  # or ti.loop_unique(x[i], covers=y.parent())
        y[p] = i


@ti.kernel
def foo():
    # x[4] = 10  # this will cause y[2] = 4 in the first test() to be incorrectly eliminated
    x[6] = 3
    x[8] = 4


@ti.kernel
def print():
    for i in y:
        print(i, y[i])


x[0] = 0
x[2] = 1
x[4] = 2
test()
foo()
test()
print()
```

g2p2g:
before: total running time(including compilation, adding cube, etc.) 66.7s, substep time 27.86ms, g2p2g time 22.38ms
after: total running time 68.6s, substep time 26.84ms, g2p2g time 21.41ms
(sync mode: total running time 76.9s, substep time 35.02ms, g2p2g time 28.73ms)

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

File: taichi/program/async_utils.cpp [hunk 1]
```diff
@@ -331,14 +331,33 @@ TaskMeta *get_task_meta(IRBank *ir_bank, const TaskLaunchRecord &t) {
...
+    // Assume we write sn on every indices we access in this task,
+    // because we would have inserted the input value state in
+    // get_meta_input_value_states otherwise.
     if (state.type == AsyncState::Type::value && state.holds_snode()) {
       const auto *sn = state.snode();
-      if (meta.element_wise.find(sn) == meta.element_wise.end() ||
-          !meta.element_wise[sn]) {
+      bool completely_overwriting = false;
+      if (meta.element_wise[sn]) {
+        // If every accesses on sn are element-wise, then it must be
```
Comment (yuanming-hu) [inline taichi/program/async_utils.cpp:343]:
```suggestion
        // If every access on sn is element-wise, then it must be
```

File: taichi/analysis/same_statements.cpp [hunk 2]
```diff
@@ -168,40 +168,65 @@ class IRNodeComparator : public IRVisitor {
...
 
-    if (check_same_value_ && stmt->is<GlobalPtrStmt>()) {
-      // Special case: we do not care the "activate" field when checking
-      // whether two global pointers share the same value.
-      // And we cannot use irpass::analysis::definitely_same_address()
-      // directly because that function does not support id_map.
+    bool field_checked = false;
+    if (check_same_value_) {
+      if (stmt->is<GlobalPtrStmt>()) {
+        // Special case: we do not care the "activate" field when checking
```
Comment (yuanming-hu) [inline taichi/analysis/same_statements.cpp:174]:
```suggestion
        // Special case: we do not care about the "activate" field when checking
```

File: taichi/ir/statements.cpp [hunk 3]
```diff
@@ -88,6 +88,18 @@ bool GlobalPtrStmt::is_element_wise(SNode *snode) const {
...
+bool GlobalPtrStmt::covers_snode(const SNode *snode) const {
+  // Check if the addresses of this statement all over the loop cover
+  // all active indices of the snode.
+  for (auto &index : indices) {
+    if (auto loop_unique = index->cast<LoopUniqueStmt>()) {
+      if (loop_unique->covers_snode(snode))
+        return true;
+    }
+  }
+  return is_element_wise(snode);
```
Comment (yuanming-hu) [inline taichi/ir/statements.cpp:100]: I'm not entirely sure what `covers` means when the SNode has multiple indices:
```python
for i, j in x:
    p = ti.loop_unique(x[i, j], covers=z.parent())
    y[i, p] = 0 # Is `y` covered by all pairs (i, p)? According to the code here it seems so.
``` 

I think either behavior is fine - or we can limit all inputs to `covers` to be 1D fields.

File: taichi/analysis/same_statements.cpp [hunk 4]
```diff
@@ -168,40 +168,65 @@ class IRNodeComparator : public IRVisitor {
...
     }
 
-    // operand check
-    if (stmt->num_operands() != other->num_operands()) {
-      same = false;
-      return;
+    bool operand_checked = false;
+    if (check_same_value_) {
+      if (stmt->is<RangeAssumptionStmt>()) {
+        // Special case: we do not care the "base" operand when checking
```
Comment (yuanming-hu) [inline taichi/analysis/same_statements.cpp:208]:
```suggestion
        // Special case: we do not care about the "base" operand when checking
```

**Review by yuanming-hu — APPROVED** [review]
Cool! LGTM. Thanks!

File: taichi/ir/statements.cpp (hunk 3 above)
Comment (xumingkuan) [reply taichi/ir/statements.cpp:100]: Yes, `y` is covered by all pairs (i, p) according to the code here.

I think we can limit all inputs to `covers` to be 1D fields.

**Pull request merged**